interfaces: add accel interface

frenchwr opened the pull request:
This adds a simple interface for accessing char device nodes managed by the relatively new [accel kernel subsystem](https://docs.kernel.org/accel/introduction.html). This subsystem is in active use by Intel NPUs (neural processing units) and Intel Gaudi SoCs, and is expected to be adopted by other vendors in the future as well.

I've tested this new interface with a validation app that ships with the [`intel-npu-driver` snap](https://github.com/canonical/intel-npu-driver-snap) and it works as expected. Currently this snap (and a handful of others that our team manages for Intel) relies on the `custom-device` interface for accessing the accel device nodes, which also works fine, but a dedicated interface will be simpler to work with for snap creators and maintainers, especially as usage of the accel subsystem becomes more widespread.

zyga commented:
What kind of hardware is required to test this? I have a core ultra 1 (mobile) and core ultra 2 (desktop) available.

zyga commented on interfaces/builtin/accel.go:
```diff
@@ -0,0 +1,56 @@
+#
+# https://docs.kernel.org/accel/introduction.html
+#
+/dev/accel/accel* rw,
```
On my system this is owned by the `render` group. The default user, at least on Fedora, is not a member of that group. This is just a statement, no action is necessary here.

zyga reviewed (approved):
First pass, looks okay but I wonder how it fits into the larger work around hardware accelerators.

frenchwr commented:
I see one of the CI jobs failed with:

```
The following interfaces do not have a documentation webpage:  accel
```

If there are some additional docs that should be added some place I'm happy to do so. I grep'd around a bit to search for docs for existing interfaces and didn't see obvious place. :)

bboozzoo commented on interfaces/builtin/accel.go:
```diff
@@ -0,0 +1,56 @@
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
```
I'm not sure if we need to deny auto connect, if we allow one for opengl through which most compute devices have been accessed until now. @pedronis wdyt?

frenchwr replied:
Shall I go ahead and push that change or is it better to wait for @pedronis to chime in?

bboozzoo replied:
Yeah, I think you can change the declaration to match the one of the opengl interface. We can always revert the patch if required.

pedronis replied:
given that it seems this  is protected by a group I'm not sure what was said about being the same as opengl applies

frenchwr replied:
> given that it seems this is protected by a group I'm not sure what was said about being the same as opengl applies

@pedronis Can you help me understand how this is relevant to the auto connection question?

FWIW, the `opengl` interface contains some support for Intel GPUs and OpenCL (which is required for supporting most Intel GPUs) yet the render device nodes for Intel GPUs (e.g. `/dev/dri/renderD128`) are also owned by the `render` group (the same group that owns the `accel` nodes for Intel NPU and Gaudi) and have group read and write permissions enabled.

jslarraz replied:
From the security point of view I think @zyga is right, if multiple processes can open this without spying on each other or tampering with each other input, then I don't see a reason to deny the auto-connection. 

My only question is whether it is the case or not (bugs aside). Whilst I hope so, I have no experience with compute devices and I could found evidences of that out there.

frenchwr replied:
I received a reply from one of the maintainers (Jozef Wludzik) of the [Intel NPU driver](https://github.com/intel/linux-npu-driver/) and he confirmed that spying is not possible:

> No, it is not possible. The memory allocation has a context isolation in NPU, which means that every zeContext has its own NPU address space, this causes that every processes uses its own NPU address space isolating access from each other. There is no possibility to read other user or other process memory without privilege access

Based on this I will go ahead and update `deny-auto-connection` to `false`.

frenchwr replied:
Done. I also went ahead and rebased my fork so everything is now up to date. @jslarraz @zyga @pedronis @bboozzoo

frenchwr commented:
@bboozzoo @zyga @alfonsosanchezbeato @jslarraz Hello, just checking in on the status of this PR. Anything I can do to help push this along?

zyga reviewed (approved):
LGTM, shall we merge this for 2.71?

frenchwr commented:
> Unit tests need an update:
> 
> ```
> ----------------------------------------------------------------------
> FAIL: basedeclaration_test.go:140: baseDeclSuite.TestAutoConnection
> 
> basedeclaration_test.go:200:
>     c.Check(err, NotNil, comm)
> ... value = nil
> ... accel
> 
> OOPS: 92 passed, 1 FAILED
> --- FAIL: TestPolicy (0.13s)
> FAIL
> coverage: 93.5% of statements
> FAIL	github.com/snapcore/snapd/interfaces/policy	0.151s
> ```

Fixed and rebased.

 I see the `unit tests (C gcc)` failed but it looks like what I assume was a transient issue with installing Debian dependencies.

pedronis reviewed (commented):
small comment

bboozzoo reviewed (approved):
LGTM

ernestl commented on interfaces/builtin/accel.go:
```diff
@@ -0,0 +1,56 @@
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: false
```
confirmed resolved

ernestl reviewed (commented):
See noble test failure relating to interface docs:
We need 'accel' interface documentation OR exclusion in the test to allow for more time to get this in place. 

I will arrange for this on the documentation side. 

Short term: @frenchwr do you mind [adding exclusion to the test](https://github.com/canonical/snapd/blob/master/tests/main/document-interfaces-url/task.yaml#L47)
Medium term: @frenchwr it will probably make sense for you to create a draft for @degville to work from.

degville commented:
> @degville, will help us out with a shell doc page, in progress...

I've created a very brief interface page that will need some expansion: https://forum.snapcraft.io/t/the-accel-interface/47396

This is now being published here, which should mean the tests now pass: https://snapcraft.io/docs/accel-interface
